aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,394 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+//! Key Transport: KEMs, including support for Kyber.
+//!
+//! # Example
+//!
+//! Note that this example uses Kyber-512, but other algorithms can be used
+//! exactly the same way, just substituting
+//! `KemAlgorithm::<desired_algorithm_here>` for `KemAlgorithm::KYBER512_R3`.
+//!
+//! ```
+//! use aws_lc_rs::key_transport::{KemAlgorithm, KemPrivateKey, KemPublicKey};
+//!
+//! let priv_key = KemPrivateKey::generate(KemAlgorithm::KYBER512_R3)?;
+//!
+//! // Generate private key bytes to possibly save for later decapsulation
+//! let privkey_raw_bytes = priv_key.as_ref();
+//!
+//! let pub_key = priv_key.compute_public_key()?;
+//!
+//! // Get the public key bytes to send to bob through some encoded
+//! // protocol message.
+//! let pub_key_bytes = pub_key.as_ref();
+//!
+//! let mut ciphertext: Vec<u8> = vec![];
+//!
+//! let retrieved_pub_key = KemPublicKey::new(KemAlgorithm::KYBER512_R3, pub_key_bytes)?;
+//! let bob_result = retrieved_pub_key.encapsulate(|ct, ss| {
+//!     ciphertext.extend_from_slice(ct);
+//!     // In a real application, we'd apply a KDF to the shared secret and the
+//!     // public keys (as recommended in RFC 7748) and then derive session
+//!     // keys from the result. We omit all that here.
+//!     Ok(())
+//! });
+//!
+//! // Retrieve private key from stored raw bytes
+//! let retrieved_priv_key = KemPrivateKey::new(KemAlgorithm::KYBER512_R3, privkey_raw_bytes)?;
+//!
+//! let alice_result = retrieved_priv_key.decapsulate(&mut ciphertext, |ss| {
+//!     // In a real application, we'd apply a KDF to the shared secret and the
+//!     // public keys (as recommended in RFC 7748) and then derive session
+//!     // keys from the result. We omit all that here.
+//!     Ok(())
+//! });
+//!
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+use crate::{
+    error::{KeyRejected, Unspecified},
+    ptr::LcPtr,
+    ptr::Pointer,
+};
+use aws_lc::{
+    EVP_PKEY_CTX_kem_set_params, EVP_PKEY_CTX_new, EVP_PKEY_CTX_new_id, EVP_PKEY_decapsulate,
+    EVP_PKEY_encapsulate, EVP_PKEY_get_raw_private_key, EVP_PKEY_get_raw_public_key,
+    EVP_PKEY_kem_new_raw_public_key, EVP_PKEY_kem_new_raw_secret_key, EVP_PKEY_keygen,
+    EVP_PKEY_keygen_init, EVP_PKEY, EVP_PKEY_KEM, NID_KYBER512_R3,
+};
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+use zeroize::Zeroize;
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Debug, PartialEq)]
+/// A selection of algorithms to be used as KEMs.
+pub enum KemAlgorithm {
+    /// NIST Round 3 iteration of the Kyber-512 algorithm
+    KYBER512_R3,
+}
```

Comment:
* Any public `enum` like should also be `#[non_exhaustive]`.
* It might be better for `KemAlgorithm` to be a pub struct (with fields for those constants) for which we provide `static` references. This is how we expose algorithms in the other modules: [aead](https://docs.rs/aws-lc-rs/1.2.0/aws_lc_rs/aead/index.html#statics), [hkdf](https://docs.rs/aws-lc-rs/1.2.0/aws_lc_rs/hkdf/index.html#statics), [cipher](https://docs.rs/aws-lc-rs/1.2.0/aws_lc_rs/cipher/index.html#statics), ...